README.md
# eth-wallet 

Command line wallet for Ethereum and ERC20 tokens.

## Installation
1. Clone this repository  
2. Optionally set virtual environment  
3. Navigate to the project's root dir and run `pip3 install .`  
4. Run `eth-wallet --help` to see if wallet was installed successfully  
 
## Usage

```
eth-wallet --help
Usage: eth-wallet [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  add-token         Add new ERC20 contract.
  get-balance       Get address balance.
  get-wallet        Get wallet account from encrypted keystore.
  list-tokens       List all added tokens.
  network           Get connected network (Mainnet, Ropsten) defined in...
  new-wallet        Creates new wallet and store encrypted keystore file.
  restore-wallet    Creates new wallet and store encrypted keystore file.
  reveal-seed       Reveals private key from encrypted keystore.
  send-transaction  Sends transaction.
```

### Create wallet
Create new wallet:
```
$ eth-wallet new-wallet   
  Passphrase from keystore: 

Account address: 0xB1f761734F00d1D368Ce6f82F755bBb3005538EB
Account pub key: 0xf94e03524a1bd803ee583a1f0de7eb1eb67a90d6802eeac22b90cfdd7ff491039441472e8db543467c0450d1b7c31b5e8f81616b99226775770f9dd531afd31a
Keystore path: /Users/Joe/.eth-wallet/keystore
Remember these words to restore eth-wallet: omit speak giant bright enable increase tube worth object timber bleak bullet
```
Show wallet:
```
$ eth-wallet get-wallet   
Account address: 0xB1f761734F00d1D368Ce6f82F755bBb3005538EB
Account pub key: 0xf94e03524a1bd803ee583a1f0de7eb1eb67a90d6802eeac22b90cfdd7ff491039441472e8db543467c0450d1b7c31b5e8f81616b99226775770f9dd531afd31a
```

### Balances
Get ETH wallet balance:
```
$ eth-wallet get-balance
Balance on address 0xB1f761734F00d1D368Ce6f82F755bBb3005538EB is: 1.234ETH
```
Add new ERC20 contract:
```
$ eth-wallet add-token
  Contract address []: 0x70a68593BAfc497AC4F24Eaf13CF68E74135bA42
  Token symbol []: ZRX
  
New coin was added! ZRX 0x70a68593BAfc497AC4F24Eaf13CF68E74135bA42
```
Get balance of ERC20 token:
```
$ eth-wallet get-balance --token ZRX
Balance on address 0xB1f761734F00d1D368Ce6f82F755bBb3005538EB is: 0.0ZRX
```

### Transactions
Send ether to another wallet
```
$ eth-wallet send-transaction 
  To address: []: 0xAAD533eb7Fe7F2657960AC7703F87E10c73ae73b
  Value to send: []: 0.01
  Password from keystore: 

transaction: {'to': '0xAAD533eb7Fe7F2657960AC7703F87E10c73ae73b', 'value': 10000000000000000, 'gas': 21000, 'gasPrice': 20000000000, 'nonce': 0, 'chainId': 3}
Pending.................
Transaction mined!
Hash of the transaction: 0x193919d1ad2dc024349ccc035a15a697987bd33e1ff04e33f878e6f89f2ebbdf
Transaction cost was: 0.00042ETH
```

Send ERC20 contract tokens to another wallet
```
$ eth-wallet send-transaction --token FIT
  To address: []: 0xAAD533eb7Fe7F2657960AC7703F87E10c73ae73b
  Value to send: []: 0.9
  Password from keystore:
 
transaction: {'to': '0x19896cB57Bc5B4cb92dbC7D389DBa6290AF505Ce', 'value': 0, 'gas': 36536, 'gasPrice': 20000000000, 'nonce': 2, 'chainId': 3, 'data': '0xa9059cbb000000000000000000000000aad533eb7fe7f2657960ac7703f87e10c73ae73b0000000000000000000000000000000000000000000000000c7d713b49da0000'}
Pending......................
Transaction mined!
Hash of the transaction: 0x118556d192c2efb13ade6ccc2f18a631e14256972af9f7ec8a67067aaafc978c
Transaction cost was: 0.00073072ETH
```

### Wallet utils
Show connected network:
```
$ eth-wallet network                
You are connected to the Ropsten network!
```
List all added tokens:
```
$ eth-wallet list-tokens
ETH
ZRX
```
Restore wallet:
```
$ eth-wallet restore-wallet
  Mnemonic sentence []: omit speak giant bright enable increase tube worth object timber bleak bullet
  Passphrase:
   
Account address: 0xB1f761734F00d1D368Ce6f82F755bBb3005538EB
Account pub key: 0xf94e03524a1bd803ee583a1f0de7eb1eb67a90d6802eeac22b90cfdd7ff491039441472e8db543467c0450d1b7c31b5e8f81616b99226775770f9dd531afd31a
Keystore path: /Users/Joe/.eth-wallet/keystore
Remember these words to restore eth-wallet: omit speak giant bright enable increase tube worth object timber bleak bullet
```
> Mnemonic sentence isn't fully compatible with BIP32 and BIP39 wallets. Therefore, only this implementation can reproduce mnemonic sentence and recreate seed!

Reveal wallet master private key:
```
$ eth-wallet reveal-seed   
  Password from keystore: 
  
Account prv key: 0x843844a23e3ae7b6a695a346c981484b554ff1718299b0b42df3045f04b94f05
```

## GUI

For the testing purpose, an additionally GUI is available. 
However, it is not recommended to use since GUI doesn't provide full functionality.
Run `eth_wallet/ui/gui.py` to launch the GUI application.
#### Example
![Alt text](doc/imgs/mac-home-page.png?raw=true "Wallet's home page!")  



CHANGELOG.md
# 1.1.2

* Update README

# 1.1.1

* Simple GUI for testing purpose
* Fixes

# 1.1.0

* Add support for ERC20 smart contract manipulation and transactions
* Gas estimation for sending ERC20 tokens
* Support of restore wallet with mnemonic sentence (BIP39)
* It is possible to check on what network is wallet connected (Mainnet/Ropsten/...)
* Readme update with commands to work control wallet

# 1.0.0

* First working wallet with one account
* Sending transactions
* Reading balance

# 0.0.1

* Command tool for key manipulation
* Key storing in single keystore


eth_wallet/contract.py
import time
from eth_wallet.infura import (
    Infura,
)
from eth_wallet.utils import (
    get_abi_json,
)


class Contract:
    """Abstraction over ERC20 tokens"""

    # fitcoin_address = '0x19896cB57Bc5B4cb92dbC7D389DBa6290AF505Ce'
    # binancecoin_address = '0x64BBF67A8251F7482330C33E65b08B835125e018'
    # my_address = '0xc3519C4560BcfE3Ac0b137f1067d1655ed65FEa4'
    # metamask_address = '0xAAD533eb7Fe7F2657960AC7703F87E10c73ae73b'

    def __init__(self, configuration, address):
        """
        Constructor
        :param address: contract address or ESN name
        :type address: string
        """
        self.conf = configuration
        self.address = address

        self.w3 = Infura().get_web3()
        self.contract = self.w3.eth.contract(address=address, abi=get_abi_json())
        self.contract_decimals = self.contract.functions.decimals().call()

    def add_new_contract(self, contract_symbol, contract_address):
        """
        Add ERC20 token to the wallet
        :param contract_symbol: token symbol
        :param contract_address: contract address
        :return:
        """
        self.conf.add_contract_token(contract_symbol, contract_address)

    def get_balance(self, wallet_address):
        """
        Get wallet's ballance of self.contract
        :param wallet_address: this wallet address
        :type wallet_address: string
        :return: balance as decimal number
        """
        return self.contract.functions.balanceOf(wallet_address).call() / (10 ** self.contract_decimals)

    def get_decimals(self):
        """
        Returns the number of decimals
        :return: integer
        """
        return self.contract_decimals

    def get_erc20_contract(self):
        """
        Returns w3.eth.contract instance
        :return:
        """
        return self.contract


# w3 = Infura().get_web3()
#
# contract = w3.eth.contract(address=fitcoin_address, abi=get_abi_json())
# # contract = w3.eth.contract(address=binancecoin_address, abi=get_abi_json())
# print(contract.functions.balanceOf(my_address).call() / 10**18)
# print(contract.functions.decimals().call())

#
# nonce = w3.eth.getTransactionCount(my_address)
# txn_dict = contract.functions.transfer(metamask_address, 2 * (10**18)).buildTransaction({
#     'chainId': 3,  # Ropsten
#     'gas': 140000,
#     'gasPrice': w3.toWei('40', 'gwei'),
#     # 'gasPrice': w3.eth.gasPrice * 10 * 2,
#     'nonce': nonce,
# })
#
# signed_txn = w3.eth.account.signTransaction(txn_dict, private_key=priv_key)
# tx_hash = w3.eth.sendRawTransaction(signed_txn.rawTransaction)
#
# print('Pending', end='', flush=True)
# while True:
#     tx_receipt = w3.eth.getTransactionReceipt(tx_hash)
#     if tx_receipt is None:
#         print('.', end='', flush=True)
#         time.sleep(1)
#     else:
#         print('\nTransaction mined!')
#         break
# print(tx_hash.hex())

